### ISSUE TITLE
WebGPU render target?

### ISSUE THREAD

**Issue by SamuelMarks** [tinygrad/tinygrad#1030]
Curiously with my GPU neither TensorFlow nor PyTorch work on Linux|Windows, but WebGPU with Chrome on Windows does. Will probably contribute to ROCm to add support, but in the interim it seems like a handy target for a project like tinygrad.

Of all tutorials, I found this one to be cleanest - https://webgpufundamentals.org/webgpu/lessons/webgpu-fundamentals.html

But I suppose the question becomes how to target WebGPU. Would one generate the JavaScript code directly, or with CPython soon getting WASM support should this be contributed to CPython and then tinygrad used to test this support on that CPython? - Or target C, as you do, and transform the output for WebGPU support (e.g., I could extend my cdd-c compiler project to support that; or could use LLVM's LibTooling).

PS: [offtopic] Oh also if you're interested in this support, maybe I'll enhance my (Apache-2.0 OR MIT) cdd-python compiler project to translate entire codebases to tinygrad. I'm thinking of taking PyTorch Lightning, torchvision and similar projects and auto-translating their calls to be tinygrad calls; and their imports to tinygrads'. Would be interesting from a regression testing perspective, but it should be doable without causing digressions from their master branches. I.e., you could automatically get all their loss functions, optimisers, models, data loaders &etc. [/offtopic]

Anyway; I digress. What do you/community think about WebGPU? - A good prospective target, or something to rely on other target backends to support?

**Comment by geohot** [comment]
I hate the 17 transpiliers in everything these days.

If you want to add WebGPU support, it would be similar to the efficientnet C support, and would output one clean JS file that runs the model that pulls in no libraries.

**Comment by dc-dc-dc** [comment]
There are some node packages that expose webdawn bindings, can use that to avoid needing a headless browser. But you would need to generate the WGSL shaders and then the js code that creates the buffers and pipeline etc.

**Comment by geohot** [comment]
node would be cool. Instead of the static webpage, I think the idea is a backend that spawns a long running node process where all the WebGPU stuff actually lives.

**Comment by SamuelMarks** [comment]
Yeah everything is a transpiler. This converts into that which converts once more. I remember complaining along with an old assembly programmer about why operating systems are so slow considering the modern state of computer hardware. Probably if we were to write everything in the intended level of abstraction we would achieve remarkable optimisations.

---

The project I found rather useful; that I was referencing in my first post; is: https://mlc.ai/web-llm/

---

Including an email I sent @geohot for context:
> I've written a bunch of open-source compilers—source-to-source, a.k.a., transpilers—in and for Rust, Swift, Java, Kotlin, C, [TypeScript, ] and Python.
> 
> Coming across your project and I was thinking, why not take someones big SDK and convert calls to be compatible with yours; and imports?
> 
> Wouldn't be too difficult in Python with my cdd-python tool (or asottile's pyupgrade + tokenise-rt).
> 
> PS: As for WebGPU, you could do it the same way as your existing backends, or use my cdd-c tool—or ROSE or LLVM's LibTooling—to convert the C code produced for clang to that target.
> 
> The advantage to this approach for all these problems is that it will automatically modify, and importantly for the first concept: allow you to keep up-to-date with the features of big projects (without necessitating becoming a big project yourself).
> 
> The disadvantages are twofold: time to develop the compilers to support these use-cases; and maintaining said compilers. Basically all the complexity of maintaining tinygrad longterm goes into the transformation library.
> 
> Anyway: enough pitching! - Would welcome your thoughts,

---

[webdawn](https://github.com/hexops/dawn) is interesting. I like how it's simple C code in a header, it would be trivial to wrap this in any language.

As for how others have implemented, here's some inspiration:
- TensorFlow.js https://github.com/tensorflow/tfjs/tree/master/tfjs-backend-webgpu/src
- JavaScript, syntax inspired-by PyTorch https://github.com/praeclarum/webgpu-torch
- [Apache TVM](https://tvm.apache.org) (the secret-sauce behind the aforementioned web-llm)
- https://github.com/milhidaka/webgpu-blas

PS: Some more nice tutorials I came across:
- https://developer.chrome.com/articles/gpu-compute/
- https://jott.live/markdown/webgpu_safari (1 year old, so YMMV with latest spec adhering clients)
- https://www.cg.tuwien.ac.at/research/publications/2023/PETER-2023-PSW/PETER-2023-PSW-.pdf (someone's undergrad thesis, pretty short and approachable)
- https://praeclarum.org/2023/05/19/webgpu-torch.html (see its github repo 6 bullet points above)

**Comment by dc-dc-dc** [comment]
Looks like they just reimplement in javascript directly

Yeah since webdawn provides a header file we can maybe look of just adding it as a runtime to execute generated WGSL shaders avoiding the need for js altogether. Then have a export script like efficientnet that generates the actual JS code to run in the browser.

**Comment by geohot** [comment]
If that header file is a stable API + webdawn is available as a Ubuntu package, I agree this is this way to go and deal with the JS on export.

**Comment by zk-tarts** [comment]
I've spent a little bit of time recently trying to make a webgpu backend work for fun.  

some things :
[wgpu-native](https://github.com/gfx-rs/wgpu-native/) has binary releases on GitHub.
As far as I know wgpu is the mozilla webgpu implementation and Dawn is the google implementation. I don't see any dawn releases anywhere.
There's a [python binding](https://github.com/pygfx/wgpu) to wgpu. It hasnt been updated to the latest version yet.

**Comment by dc-dc-dc** [comment]
Looks like [wgpu-py](https://github.com/pygfx/wgpu-py) is using them and is being kept up to date. Their api seems to match the js webgpu api so this might be the way to go. Webdawn had their own api with way more boilerplate as well as it had to be compiled which took well over 10 mins for me.

**Comment by audiovention** [comment]
Would be very excited to see wgpu backend. I can confirm I used wgpu-py for a small project, everything works very well (tested on Linux/Windows integrated AMD gpu, and mac M2). The major part of the code would be in the kernels anyway I guess, which are not dependent on the bindings used.
